query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -335,39 +433,53 @@ private String getLogicalColumnName(Expression expression) {
     }
   }
 
-  private LiteralConstant[] convertExpressionToMapLiterals(Expression expression) {
+  private LiteralConstant[] convertExpressionToMapLiterals(Expression rhs, Expression lhs) {
     LiteralConstant[] literals = new LiteralConstant[2];
-    if (expression.getValueCase() == LITERAL) {
-      LiteralConstant value = expression.getLiteral();
+    String[] literalArguments = new String[] {"", ""};
+
+    if (rhs.getValueCase() == LITERAL) {
+      LiteralConstant value = rhs.getLiteral();
+      // backward compatibility
       if (value.getValue().getValueType() == ValueType.STRING_ARRAY) {
-        for (int i = 0; i < 2 && i < value.getValue().getStringArrayCount(); i++) {
-          literals[i] =
-              LiteralConstant.newBuilder()
-                  .setValue(
-                      Value.newBuilder().setString(value.getValue().getStringArray(i)).build())
-                  .build();
+        literalArguments[0] = value.getValue().getStringArray(0);
+        if (value.getValue().getStringArrayCount() > 1) {
+          literalArguments[1] = value.getValue().getStringArray(1);
         }
+      } else if (value.getValue().getValueType() == ValueType.STRING) {
+        literalArguments[0] = lhs.getAttributeExpression().getSubpath();
+        literalArguments[1] = value.getValue().getString();
       } else {
         throw new IllegalArgumentException(
             "operator CONTAINS_KEYVALUE supports "
                 + ValueType.STRING_ARRAY.name()
+                + " and "
+                + ValueType.STRING.name()
                 + " value type only");
       }
     }
 
-    for (int i = 0; i < literals.length; i++) {
-      if (literals[i] == null) {
-        literals[i] =
-            LiteralConstant.newBuilder().setValue(Value.newBuilder().setString("").build()).build();
-      }
+    for (int i = 0; i < 2; i++) {
+      literals[i] =
+          LiteralConstant.newBuilder()
+              .setValue(Value.newBuilder().setString(literalArguments[i]).build())
+              .build();
     }
-
     return literals;
   }
 
-  private boolean isComplexAttribute(Expression expression) {
-    return expression.getValueCase().equals(ATTRIBUTE_EXPRESSION)
-        && expression.getAttributeExpression().hasSubpath();
+  private boolean isColumnMapAttribute(Expression expression) {
+    return expression.getValueCase() == COLUMNIDENTIFIER
```

Comment:
I thought this would be easy to tackle since we eventually have to deprecate the use of column identifier in future. Also combining them into a single function was leading to addition of extra conditions in the places where they are used.

Reply:
> I thought this would be easy to tackle since we eventually have to deprecate the use of column identifier in future. Also combining them into a single function was leading to addition of extra conditions in the places where they are used.

Not sure I follow here - we should never care in the calling code if the same information is passed in via a column identifier or an attribute expression (hence my suggestion to preprocess and normalize all column identifiers as attribute expressions, which makes it really easy to drop column identifiers in the future since none of the transform code would depend on it). This may be a case of terminology as discussed above though.

Reply:
renamed accordingly

Reply:
resolved